### ISSUE TITLE
jax.value_and_grad return gradient is zero when use jax.random.categotical

### ISSUE THREAD

**Issue by WangHB0** [google/jax#17556]
### Description

I want to use `jax.random.categorical` to sampling from a probability matrix generated via `flax.linen.Dense`. And use `jax.value_and_grad` to calculate the gradient between samples and my costmize target named `ground`. But when I use `jax.random.categorical` to sampling, the gradient is zero. While just use probability matrix and target to calculate the gradient, it is nonzero. 

    import jax
    import optax
    import flax.linen as nn
    import jax.numpy as jnp
    from flax.training import train_state

    def calculate_loss(params, model, trg, input):
        prob_list = model.apply(params, input)
        samples = jax.random.categorical(sample_rng, prob_list, axis=1)
        loss = jnp.mean(jnp.square(samples - trg))

        # trg = jnp.repeat(jnp.array([trg]), repeats=prob_list.shape[1], axis=1)
        # loss = jnp.mean(jnp.square(prob_list - trg))
        return loss

    layer = nn.Sequential([nn.Dense(4), nn.softmax])
    main_rng = jax.random.PRNGKey(0)
    main_rng, sample_rng, dense_rng = jax.random.split(main_rng, 3)

    init_sample = jnp.array([[1,1,0,0]])
    params = layer.init(dense_rng, init_sample)

    optimizer = optax.adam(learning_rate=0.1)
    state = train_state.TrainState.create(apply_fn=layer.apply, params=params, tx=optimizer)

    ground = jnp.array([0])
    loss = lambda params: calculate_loss(params, layer, ground, init_sample)
    loss_value, gradient = jax.value_and_grad(loss)(state.params)
    state = state.apply_gradients(grads=gradient)
    print(gradient)

### What jax/jaxlib version are you using?

0.4.13

### Which accelerator(s) are you using?

CPU

### Additional system info

python version: 3.9

**Comment by jjyyxx** [comment]
It works as intended. `jax.random.categorical` returns indicies of integral type `jnp.int_`. It is impossible to calculate gradient with such values. Typically, optimizing a distribution against a label involves calculating `log_prob` of `target` with the distribution, with packages like `numpyro` (or hand-written), instead of sampling.

**Comment by WangHB0** [comment]
Great, thank you very much for your reply, I will try using `log_prob` of `target` to optimize

**Comment by WangHB0** [comment]
hello @jjyyxx 
I have thought carefully about the method you proposed. Do you mean use network to calculate `log_prob` of `target` and `log_prob` of `samples` (parameters of network is same). Then calculate the `loss` with their `log_prob`. Finally, use `loss` to optimize.

**Comment by whopezx** [comment]
In fact, when you sample, it is equivalent to applying a _**step function**_ to this data set, so the gradient of result obtained must be 0. If you want to evaluate whether your sample meets your requirements, you need to input the sample into your model and use the model's output for evaluation. If there is only a sample, it can only be judged through some physical or statistical knowledge.
